### PR TITLE
Using JSON to import comsol results 

### LINKED ISSUE #4026: Decide how to bundle `.pickle` files from COMSOL solution results data
`.pickle` files are binary files and not inherently secure. They therefore should be best included elsewhere in PyBaMM's packaging infrastructure when publishing to common distribution channels like PyPI owing to security reasons in the case where these binaries are modified with malicious intent and make it to the `develop` branch or a PyPI release where PyBaMM can be downloaded from (binary or non-binary). This means they should neither be included in the source distribution (controlled by `MANIFEST.in`), nor in the wheels (in the case where the wheel is built from the sdist). Binary platform-specific wheels for PyBaMM cannot be built from sdists right now and have to be built in-tree (#3651, #3564), and require restructuring the build system configuration at least, if not rewriting what files are copied into the package structure – since `pybamm/input/comsol_results/` is a namespace package (via `setuptools.command.install_data`).

We have the following options:

1. Store in JSON, CSV, or TXT file formats (non-binary) instead of `.pickle`
2. Use [`pooch`](https://learn.scientific-python.org/development/patterns/data-files/) to reproducibly download these files when needed and adjust PyBaMM's public API plus example notebooks for this, while moving these files to a separate repository in the `pybamm-team` organisation, and marking the repository as an archive to prevent tampering.
3. Use non-binary storage solutions from `xarray` or `zarr` because they can handle multi-dimensional data

Out of these, point 2 requires an internet connection and those wishing to load COMSOL results from these files would need access to the internet to download them, and therefore should be the last option to consider.

<hr>

## Task list, as of 10/05/2024

There are two major things to be done:

- [x] Convert pickle files to non-binary files (done in #4064)
- [ ] Use https://github.com/fatiando/pooch as an optional (`[dev]`) dependency to fetch these files plus adjust for the change here
    - [x] Set up a new repository: https://github.com/pybamm-team/pybamm-data
        - [ ] Add a README and some documentation
        - [ ] Copy the files and delete them from the PyBaMM main (this) repository (and additionally store checksums in a text file somewhere), and publish a GitHub Release there to store them perhaps (shorter URL)
    - [ ] Refer to the Scientific Python guide: https://learn.scientific-python.org/development/patterns/data-files/#downloading-larger-files-on-demand
    - [ ] Use `pytest`'s conditional markers or regular markers to skip tests where access to the internet is not available (adjust `noxfile.py`, test-related files, etc.)
    - [ ] Ensure files are not copied into the sdist (control `MANIFEST.in` and `setuptools`-specific package data)
    - [ ] Unsure about this point – maybe provide a public API to fetch these files (with the URLs kept hardcoded)?

### PR DESCRIPTION
**Pull request by santacodes**
# Description

```python
    comsol_variables = json.load(open(comsol_results_path, "rb"))
    variable = np.array(comsol_variables["variable_name"])
    
```
The default imported data from JSON has list datatype so a bit of pre-processing is done before it can be used to plot. `np.array()` converts the multi-dim list to multi-dim numpy array.

Addresses a part of #4026 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by kratman** [comment]
If this works then I am happy with it. It looks like the lychee thing is unrelated

**Comment by kratman** [comment]
Lychee is fixed here #4065

**Comment by santacodes** [comment]
```python 
import json
import pickle
import numpy as np

files = ["comsol_01C.json","comsol_05C.json", "comsol_1C.json", "comsol_1plus1D_3C.json","comsol_2C.json", "comsol_3C.json"]


for file in files:
    with open(file.split('.')[0] + ".pickle", 'rb') as infile:
        pickleobj = pickle.load(infile)

        for i in pickleobj:

            if i == 'solution_time':
                pickleobj[i] = None
                continue
            else:
                pickleobj[i] = pickleobj[i].tolist()

    with open(file, 'w', encoding='utf-8') as outfile:
        json.dump(pickleobj, outfile, ensure_ascii=False, indent=4)
```
For reference, I am attaching the code I used to convert the pickle data with multi-dim arrays to JSON. JSON doesn't allow `ndarray` serialization so I had to convert the `multi-dim arrays` to a `multi-dim lists` to write to JSON. Only key name with `'solution_time'` had a NoneType as value and not array which is why the extra if statement is added to include the edge case.

File: MANIFEST.in [hunk 1]
```diff
@@ -1,5 +1,5 @@
 graft pybamm
-include CITATION.cff
+include CITATION.cff pybamm/input/comsol_results/*.json
```
Comment (agriyakhetarpal) [inline MANIFEST.in:2]: Might be better to use something like `global-include <pattern to include>`, i.e., `global-include comsol_*.json` in a new line

File: .gitignore [hunk 2]
```diff
@@ -10,14 +10,15 @@
...
-# don't ignore important .txt and .csv files
+# don't ignore important .txt, .csv and json files
 !requirements*
 !LICENSE.txt
 !CMakeLists.txt
 !pybamm/CITATIONS.bib
 !pybamm/input/**/*.csv
 !tests/unit/test_parameters/*.csv
 !benchmarks/benchmark_images/*.png
+!pybamm/input/comsol_results/*.json
```
Comment (agriyakhetarpal) [inline .gitignore:21]: The folder is probably already not-ignored by the lines

```gitignore
!input/comsol_results/
```

present below, would you be able to refine this file for redundant rules?

**Review by agriyakhetarpal — COMMENTED** [review]
Thanks, @santacodes! The Windows failure is unrelated, I just triggered a re-run for it. I haven't verified all of the JSON files, but I did for one or two of them (one can't see the diffs on GitHub – they are quite big!) on my machine, and after the conversion to arrays, the elements seem to be matching one-for-one (this should be better than the previous iteration where I felt there were missing values due to the presence of `...` in the strings) which is great.

For the example notebooks tests, it looks like there is [one more file](https://github.com/pybamm-team/PyBaMM/actions/runs/9007651749/job/24748881175?pr=4064#step:9:623) that you need to fix :) The changes look quite good to me.

**Review by valentinsulzer — COMMENTED** [review]
Looks good to me other than Agriya's changes. Would be good to keep the COMSOL results in a separate location e.g. using pooch as discussed on Slack. That can be a separate PR

**Comment by agriyakhetarpal** [comment]
> Would be good to keep the COMSOL results in a separate location e.g. using pooch as discussed on Slack. That can be a separate PR

The only trouble that I feel this will cause is that it would require an active internet connection for those who would wish to use such data for their experiments, but then for someone to download and install PyBaMM, doing that requires an internet connection anyway... so I agree that this can be done overall – @santacodes, would you be willing to do this further? I just created a new repository for this at https://github.com/pybamm-team/pybamm-data, and named it as such rather than giving a COMSOL-specific name because we can move other CSV files for drive cycles, Enertech cells, etc. as well. I'll copy the files over there once this PR gets merged.

**Comment by agriyakhetarpal** [comment]
In the absence of a network connection, we will have to skip relevant tests that use these data files. The `scikit-image` test suite has examples of this behaviour and `pytest` can be configured with fixtures, i.e., `@pytest.mark.network` and then configuring the test suite to run `pytest -m "not network"`

**Comment by santacodes** [comment]
> > Would be good to keep the COMSOL results in a separate location e.g. using pooch as discussed on Slack. That can be a separate PR
> 
> The only trouble that I feel this will cause is that it would require an active internet connection for those who would wish to use such data for their experiments, but then for someone to download and install PyBaMM, doing that requires an internet connection anyway... so I agree that this can be done overall – @santacodes, would you be willing to do this further? I just created a new repository for this at https://github.com/pybamm-team/pybamm-data, and named it as such rather than giving a COMSOL-specific name because we can move other CSV files for drive cycles, Enertech cells, etc. as well. I'll copy the files over there once this PR gets merged.

Sure I am willing to work on it, I'll get started once this PR is merged.

**Comment by kratman** [comment]
> In the absence of a network connection, we will have to skip relevant tests that use these data files.

Just an aside, but I am not able to use nox for testing without connection anyway. I think if the files were handled in a way that they are stored elsewhere and downloaded only if they don't exist would solve most issues for that

File: .gitignore (hunk 2 above)
Comment (santacodes) [reply .gitignore:21]: Sure, working on cleaning it up!

**Comment by agriyakhetarpal** [comment]
> Just an aside, but I am not able to use nox for testing without connection anyway.

This is because `nox` is installing the dependencies for you besides running the tests, so if you have your `pip` downloads cached and your virtual environments in place in `.nox/` , then it should be able to function without an internet connection. You can also do `nox -s unit --no-install` to skip the installation step that requires internet access, that will start the test runner without executing any `session.install("<...>")` statements.

>  I think if the files were handled in a way that they are stored elsewhere and downloaded only if they don't exist would solve most issues for that

Yes, `pooch` will handle this. It will store these files in your OS cache (or a specified location if needed, but this is not recommended) and will raise an error if those files are tampered with (by verifying the SHA-256 checksum for the files). It will re-download the files only if they don't exist.

**Comment by santacodes** [comment]
`pouch_cell_model.ipynb` was failing doctests which seems like an unrelated issue. I left out `manifest.in` as we would be moving the JSON files to a separate repository for pooch.

**Review by agriyakhetarpal — APPROVED** [review]
Thanks, @santacodes! All of the notebook outputs look identical. Yes, leaving out changes to `MANIFEST.in` is alright because these files will be moved to a separate repository soon anyway

**Comment by agriyakhetarpal** [comment]
> Should we add [this commit](https://github.com/pybamm-team/PyBaMM/commit/b94dedf947359f86680028ee5b1c045434d24ffb) hash to `.git-blame-ignore-revs`?

I don't think we should, because these are relevant code changes and they will not spoil the git blame (the revisions are not scattered across the codebase but are rather in a specific directory).

**Comment by agriyakhetarpal** [comment]
I edited the PR description so that it does not close the linked issue, we can track the second half there. I'll edit the issue description.

**Pull request merged**